zerver/lib/user_groups.py

```diff
@@ -26,6 +32,34 @@ def access_user_group_by_id(user_group_id: int, user_profile: UserProfile) -> Us
     return user_group
 
 
+def get_user_group_from_ids(
+    user_group_ids: Sequence[int], acting_user: UserProfile
+) -> List[UserGroup]:
+    user_groups = UserGroup.objects.filter(id__in=user_group_ids, realm=acting_user.realm)
+
+    valid_group_ids = [group.id for group in user_groups]
+    invalid_group_ids = [group_id for group_id in user_group_ids if group_id not in valid_group_ids]
+    if invalid_group_ids:
+        raise JsonableError(_("Invalid user group ID: {}").format(invalid_group_ids[0]))
+
+    user_group_members = UserGroupMembership.objects.filter(
+        user_group_id__in=user_group_ids
+    ).values_list("user_profile_id", "user_group_id")
+
+    user_group_members_dict: Dict[int, Set[int]] = defaultdict(set)
+    for (user_id, group_id) in user_group_members:
+        user_group_members_dict[group_id].add(user_id)
```

Comment:
I think this membership check is only going to verify whether the current user is a direct member, right? 

We also probably need to restrict access to system groups here.

Reply:
OK, I think this is the main outstanding issue for this PR. My previous comment is incorrect, given how this function is used. It looks like this function is used to check whether the current user has the ability to access a group for the purpose of making it a member of another group. 

I think probably the rule chosen is reasonable -- basically to add group X to another group, you need to be a moderator or a member of the group. But this would be a good thing to have a #design conversation about unless we see a clear answer for what the rule should be... I don't have a clear idea of what is correct. @alya FYI.

I think we should rename this to `access_user_groups_as_potential_subgroups` or something like that.

Reply:
OK I think after talking with @alya a bit, my proposal is that we only check that this is a group that you can see exists (i.e. that you're not a guest, and they're in your realm, at least with the current model). Neither of us could think of a product reason for why it'd be bad for someone who has permission to add individual users to a group to add another group to that group (even if they have no connection to the new subgroup), which is basically what we're considering here.

Can you post this plan in #design to make sure it gets attention?  I think it's fine to change this function to have that behavior without waiting for that thread to conclude since I think it's unlikely we'll do something different, and `access_user_groups_as_potential_subgroups` seems like a fine name to use.